spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -464,3 +464,28 @@ for (Pageable p = Pageable.ofSize(25).sortBy(Sort.desc("yearBorn"), Sort.asc("na
 }
 ----
 
+== Jakarta Data Providers
+
+=== Modes
+
+Jakarta Data Providers implement the Jakarta Data API and specification requirements for one or more of the following modes.
+
+==== Standalone Mode
+
+Standalone mode means that the Jakarta Data provider is able to run on its own outside of a Jakarta EE platform or profile. Standalone Jakarta Data providers can choose whether or not to provide their own custom dependency injection leveraging `@Inject`, whether or not to provide extensions to tie into dependency injection via Jakarta Contexts and Dependency Injection (CDI) SE, whether or not to tie into the Jakarta Data Provider SPI, and whether or not to define other vendor specific ways of accessing Jakarta Data repositories. The Jakarta Data TCK includes a Standalone mode for validating Jakarta Data Providers that requires at least one of the first three of the aforementioned approaches.
+
+==== Jakarta EE Mode
+
+Jakarta EE mode means that the Jakarta Data provider is able to run within a Jakarta EE product, which is a product that provides a Jakarta EE profile or the full Jakarta EE platform, or a product that is comprised of a subset of Jakarta EE specifications. If the Jakarta EE product is running in a mode that provides Jakarta Contexts and Dependency Injection (CDI), then the Jakarta EE product coordinates CDI dependency injection for all Jakarta Data providers that it discovers via the Jakarta Data Provider SPI. Jakarta Data providers do not need to support CDI. Jakarta Data providers can continue using their own custom dependency injection models without CDI present. The Jakarta Data TCK includes a Jakarta EE mode for validating Jakarta Data Providers that leverages the `@Inject` annotation that is common to both CDI and custom dependency injection models.
```

Comment:
> then the Jakarta EE product coordinates CDI dependency injection for all Jakarta Data providers that it discovers via the Jakarta Data Provider SPI. Jakarta Data providers do not need to support CDI.

Can we make it optional?
A CDI user can do it using only CDI. So, it should not be mandatory for any vendor.

Reply:
Pushing the responsibility onto the user would be a poor design. It's a natural fit for the Jakarta EE product to provide the CDI etxension and coordinate among the registered providers, automatically determining the correct one by the entity annotation so that the correct repository is created and injected without extra work from the user.  This will offer the best usability because it requires the customer to do the least amount of work, avoids duplication (the entity annotation alone is enough to know which provider they want), and is highly intuitive.  With the following we have a great model that provides superior usability.  Let's keep it that way.

```
@Entity
public class Product {
...
}

@Repository
public interface Products extends CrudRepository<Product, Long> {
...
}

@Inject
Products products;
```

Reply:
I agree with you that it should be this way.

What I don't understand is the relation between CDI and this Data provider.

Could you show some code to express why this is so crucial? and why not use what CDI already provides to us?

Reply:
I think we keep getting tripped up and talking past each other because there are two modes for running - Standalone vs in a Jakarta EE profile and we each tend to be talking about the opposite one than the other is thinking of.  Hopefully additional language that I've been adding when addressing review comments will help.

If the Jakarta Data provider is running Standalone and so is the only provider in the system, then having a module that provides your own CDI extension that handles all `@Repository` injection points is just fine.  The Jakarta Data provider won't collide with any others on those injection points because there aren't any others. This pull should be consistent with allowing this.

If the Jakarta Data provider is running in a Jakarta EE profile where there can be multiple Jakarta Data providers and the Jakarta EE product runtime itself, then we have a collision if each provider and the Jakarta EE product runtime are all supplying a CDI extension to take over handling of all the injections points for `@Repository`.  Two meetings ago, we discussed two different idea for avoiding the collisions.  One idea was having the Jakarta Data providers only handle injection points annotated with their own special qualifier, which would resolve the collision but at the price of adding unnecessary burden onto the end user.  The other idea was that Jakarta Data providers could just register via a provider SPI with the Jakarta EE product and the Jakarta EE product runtime would coordinate them all from its single CDI extension, which is the approach covered under this pull.  This approach _does_ use what CDI provides to us, it's just the Jakarta EE product runtime that provides the CDI extension in this case, not the Jakarta Data providers all providing a bunch of competing CDI extensions for the same injection point.  We've already tried out the solution and it works quite well.  I know it's an inconvenience if JNoSQL ends up implementing the Jakarta Data provider SPI for running in Jakarta EE profile, and for running in Standalone has a separate module with its own CDI extension.  But it shouldn't be that much trouble.  The SPI is only a single interface with a few methods on it, some of which are one-liners like the provider name "JNoSQL" and supported entity annotation class `jakarta.nosql.Entity.class` and hopefully the others like creating a repository instance can just delegate to what you already have for your CDI extension.

Reply:
I got it; the primary goal of this provider is to avoid Ambiguous Exception CDI when it has multiple vendors.

```java
@Inject
Products products;
```

Given an imaginary scenario with multiple JPA providers, such as Hibernate and Eclipse Link JPA, how can this provider help you?

I mean, those providers will return the same entity on the method `supportedEntityAnnotations()`

Reply:
A single Jakarta Data provider backed by JPA should be able to cover all relational databases, so a scenario where you need to have multiple Jakarta Data providers all backed by JPA does not seem very likely/realistic.  That scenario is more of an edge case that could be covered by configuration (associating a repository type with a specific provider name) if the need for it ever arises.  The SPI does include a method to collect the provider name for this.

Reply:
We can add if it makes sense, but please, let me understand and try to help.

The scenario here is to work with Jakarta NoSQL and Jakarta Data on the same server.

Can we assume that an entity will run only in a single database?

If yes, given a scenario where I have two entities, one to Jakarta NoSQL and one to Jakarta Data:

```java
@jakarta.nosql.Entity
class AirPlane{
....
}
```


```java
@jakarta.persistence.Entity
class Car{
....
}
```

Where we'll have two repositories:

```java
@Repository
public interface Airport extends CrudRepository<AirPlane, Long> {
}

@Repository
public interface Garage extends CrudRepository<Car, Long> {
}

```

In this case, when I inject both repositories, IMHO: there is no ambiguous exception once `Airport` is provided to the Jakarta NoSQL vendor, and `Garage` should be provided to the Jakarta Data vendor.


```java
@Inject//visible only to NoSQL
Airport airport;
@Inject//visible only to Relational database
Garage garage;
```

We need to document it somehow, such as: As the provider discovers and provides a repository that can support it; otherwise, it will ignore it.

On the code side, we can return the entity from the Repository with this code:


```java
public boolean isNoSQLEntity(Class<?> type) {
    Optional<Class<?>> entity = getEntity(type);
    return entity.map(c -> c.getAnnotation(jakarta.nosql.Entity.class) != null)
            .orElse(false);
}

public boolean isRelationalEntity(Class<?> type) {
    Optional<Class<?>> entity = getEntity(type);
    return entity.map(c -> c.getAnnotation(jakarta.persistence.Entity.class) != null)
            .orElse(false);
}

private Optional<Class<?>> getEntity(Class<?> repository) {
    Type[] interfaces = repository.getGenericInterfaces();
    if (interfaces.length == 0) {
        return Optional.empty();
    }
    ParameterizedType param = (ParameterizedType) interfaces[0];
    Type[] arguments = param.getActualTypeArguments();
    if (arguments.length == 0) {
        return Optional.empty();
    }
    Type argument = arguments[0];
    if (argument instanceof Class<?> entity) {
        return Optional.of(entity);
    }
    return Optional.empty();
}
```

So, you can check the repository type and register only where it makes sense:

```java
isNoSQLEntity(Airport.class); //true
isRelationalEntity(Airport.class);//false
isNoSQLEntity(Garage.class); //false
isRelationalEntity(Garage.class);//true
```

Hope this one can help. Please, if there is any concern.

Reply:
Nice - I didn't realize they are now open to CDI support and pursuing it.  That really changes things a lot.  Here's a writeup in a separate pull #131 of some rules for CDI extensions that we could use to avoid conflicts is most cases and in edge cases cope with them.  The approach does work when I tried it out with multiple providers and no longer requires the central abstraction to coordinate.

Reply:
@njr-11 are you running Jakarta NoSQL and Data at Open Liberty?

Reply:
> @njr-11 are you running Jakarta NoSQL and Data at Open Liberty?

The goal is to do so once the spec chooses a solution for multiple providers (could be either this pull or #131) and JNoSQL has a version that aligns with it.  In the mean time we are using other providers/mock providers to try things out.  Both approaches have shown to be workable.  We just need to choose one and more forward with it.

Reply:
Nice, I like the other pull solution.

I'll give some suggestions.